src/commands/importExportEntities/entities/utils/reference.ts

```diff
@@ -0,0 +1,77 @@
+import { ImportContext, SimplifyContext } from "../../entityDefinition.js";
+
+type Params = Readonly<{
+  newId: string | undefined;
+  oldId: string;
+  entityName: string;
+}>;
+
+export const createReference = (params: Params) =>
+  params.newId ? { id: params.newId } : { external_id: `non-existent-${params.entityName}-${params.oldId}` };
+
+/**
+ * extracts top level values from import context. `string:string` pairs are left unchanged,
+ * `string:object` are simplified, with `selfId` string replacing the object as value.
+ */
+const simplifyContext = <K extends keyof ImportContext = keyof ImportContext>(
+  context: Partial<ImportContext>,
+  keys?: K[],
+) => {
+  const getSimplifiedMap = (map: ReadonlyMap<string, string | { selfId: string }>) =>
+    new Map([...map].map(([k, v]) => [k, typeof v === "string" ? v : v.selfId]));
+
+  const keysToExtract = keys ?? Object.keys(context) as K[];
+
+  return Object.fromEntries(
+    Object.entries(context)
+      .filter(([k]) => keysToExtract.includes(k as K))
+      .map(([k, v]) => [k, getSimplifiedMap(v)]),
+  ) as unknown as SimplifyContext<K>;
+};
+
+const flattenContext = <K extends keyof ImportContext = keyof ImportContext>(
+  context: SimplifyContext | SimplifyContext<K>,
+) => new Map<string, string>(Object.values(context).flatMap(c => [...c]));
+
+/**
+ * extracts specified `keys` from `ImportContext`, simplifies `string:object`
+ * pairs to `string:(string)selfId`, then merges all pairs into a single `Map<string, string>`
+ *
+ * @param context original `ImportContext` object or its partial representation
+ * @param keys keys to extract from `ImportContext` (defaults to all)
+ * @returns map of all `oldId:newId` pairs for top level entities
+ */
+export const simplifyAndFlattenContext = <K extends keyof ImportContext = keyof ImportContext>(
+  context: Partial<ImportContext>,
+  keys?: K[],
+) => flattenContext(simplifyContext(context, keys));
+
+export const transformReferences = <T extends object>(
+  object: T,
+  flatContext: Map<string, string>,
+): T => {
+  /**
+   * this merges all simplified context maps into a single map, which is subsequently used
```

Comment:
this function does not merge... it gets merged context maps as a parameter.

Reply:
also outdated from the rebase. fixed